src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -181,6 +181,9 @@ private void EnableTimer(float pollingIntervalInSeconds)
                 // notify the polling thread that the polling interval may have changed and the sleep should
                 // be recomputed
                 s_pollingThreadSleepEvent!.Set();
+
+                // reset counters before leaving this method to prevent them from being overwritten
+                s_needsResetCounterGroupEvent!.WaitOne();
```

Comment:
I believe blocking here would recreate the deadlock conditions we are trying to avoid. When running the example described in https://github.com/dotnet/runtime/issues/93175, I would expect:

Thread A - Main thread calling Threadpool.GetMaxThreads() - Acquires static constructor lock, waits on EventListenerLock
Thread B - ETW callback thread calling EnableTimer - Acquires EventListenerLock, waits on this new event
Thread C - Timer thread calling Threadpool.get_CompletedWorkItemCount() - Waits on static constructor lock, never signals the new event

In total A waits on B waits on C which waits on A again.


I see what you mean about the tests calling WriteMetric() immediately after creating the EventSource and assuming that the data would be captured. https://github.com/dotnet/runtime/blob/main/src/tests/tracing/eventcounter/eventcounter.cs#L121

While this behavior probably isn't a promise we ever intentionally wanted to make its possible other code has taken a dependency on it too and its probably not too hard to preserve it. I'm thinking now we could reset all counter types except IncrementingPollingCounter synchronously in EnableTimer() and only do the callback that initializes IncrementingPollingCounter asynchronously on the Timer thread. That looks like it would preserve as much of the existing synchronous reset behavior as we can while still addressing the deadlock. You could resume using ResetCounters() for synchronous reset behavior and then ResetCounterGroupIfNeeded() could be repurposed into something like InitIncrementingPollingCounters().

Reply:
@noahfalk, you're right, I missed that case, sorry :( 
Thank you for pointing it out!

I pushed a version with resetting only IncrementingPollingCounters on the Timer thread, I decided to put IncrementPollingCounters in a static list. I considered static CounterGroup field as alternative, but, in this case  CounterGroup._counters list can be changed within releasing s_counterGroupLock in EnableTimer method and taking it in the Timer thread.
Could you please take a look?